src/Runtime/Eval.hs

```diff
@@ -68,69 +69,58 @@ posMatches xp yp (x, y) = match xp x && match yp y
-                                      let p2  = if bx == by && bx == 1 then "The board only has one space at (1,1)." else "The board size is ("++ show bx ++ "," ++ show by ++")." in
-                                      Err $ p1 ++ p2
-                             False -> arr ! (x,y)-- good index
-
+   case (x < 1 || x > bx || y < 1 || y > by) of
+     False -> arr ! (x,y) -- good index
+     True  -> Err $ p1 ++ p2
+         where
+            p1 = "Could not access (" ++ show x ++ "," ++ show y ++ ") on the board, " ++
+              "this is not a valid space. "
+            p2 = if bx == by && bx == 1 then "The board only has one space at (1,1)."
+                                     else "The board size is ("++ show bx ++ "," ++ show by ++")."
 
 -- | Binary operation evaluation
 evalBinOp :: Op -> (Expr a) -> (Expr a) -> Eval Val
-evalBinOp Plus l r      = evalNumOp "+" (+) l r
-evalBinOp Minus l r     = evalNumOp "-" (-) l r
-evalBinOp Times l r     = evalNumOp "*" (*) l r
-evalBinOp Div l r       = evalNumOp "/" div l r
-evalBinOp Mod l r       = evalNumOp "%" mod l r
-evalBinOp Less l r      = evalCompareOpInt (<) l r
-evalBinOp Leq l r       = evalCompareOpInt (<=) l r
-evalBinOp NotEquiv l r  = evalNotEquiv l r
-evalBinOp Equiv l r     = evalEquiv l r
-evalBinOp Geq l r       = evalCompareOpInt (>=) l r
-evalBinOp Greater l r   = evalCompareOpInt (>) l r
-evalBinOp Get l r       = do
-									board <- eval l
-									pos   <- eval r
-									case (board, pos) of
-										(Vboard arr, Vt [Vi x, Vi y]) -> return $ tryUnsafeBoardAccess (x,y) arr
-										_ -> return $ Err $ "Could not access" ++ show l ++ " in " ++ show "r"
-                 -- not a great error message, but this should be caught in the typechecker anyways
+evalBinOp Plus l r     = evalNumOp "+" (+) l r
+evalBinOp Minus l r    = evalNumOp "-" (-) l r
+evalBinOp Times l r    = evalNumOp "*" (*) l r
+evalBinOp Div l r      = evalNumOp "/" div l r
+evalBinOp Mod l r      = evalNumOp "%" mod l r
+evalBinOp Less l r     = evalCompareOpInt (<) l r
+evalBinOp Leq l r      = evalCompareOpInt (<=) l r
+evalBinOp NotEquiv l r = evalNotEquiv l r
+evalBinOp Equiv l r    = evalEquiv l r
+evalBinOp Geq l r      = evalCompareOpInt (>=) l r
+evalBinOp Greater l r  = evalCompareOpInt (>) l r
+evalBinOp Get l r      = do
+   board <- eval l
+   pos   <- eval r
+   case (board, pos) of
+      (Vboard arr, Vt [Vi x, Vi y]) -> return $ tryUnsafeBoardAccess (x,y) arr
+      _ -> return $ Err $ "Could not access" ++ show l ++ " in " ++ show "r"
 
 -- | evaluates the /= operation
--- TODO redundant, but tests are in place for both this and 'evalEquiv' below.
--- For == and /= these should be merged together into a more general form,
--- and then modified with a simple 'not' to avoid this duplication of logic
-evalNotEquiv :: (Expr a) -> (Expr a) -> Eval Val
 evalNotEquiv l r = do
-                 v1 <- eval l
```

Comment:
refactor: reduce code duplication

Reply:
Thanks for making the change here! I was a bit lazy when I wrote it, even wrote what 'should' be done 😆

Reply:
Been there, done that :laughing: